multisig/src/lib.rs

```diff
@@ -0,0 +1,300 @@
+use std::collections::HashSet;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use near_sdk::collections::Map;
+use near_sdk::json_types::{Base58PublicKey, U128};
+use near_sdk::{env, near_bindgen, AccountId, Balance, Gas, Promise, PublicKey};
+
+const DEFAULT_ALLOWANCE: u128 = 1_000_000_000_000_000_000_000;
+
+pub type RequestId = u32;
+
+#[derive(BorshDeserialize, BorshSerialize)]
+pub enum MultiSigRequest {
+    Transfer {
+        account_id: AccountId,
+        amount: Balance,
+    },
+    AddKey {
+        public_key: PublicKey,
+    },
+    FunctionCall {
+        contract_id: AccountId,
+        method_name: String,
+        args: Vec<u8>,
+        deposit: Balance,
+        gas: Gas,
+    },
+}
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct MultiSigContract {
+    num_confirmations: u32,
+    request_nonce: RequestId,
+    requests: Map<RequestId, MultiSigRequest>,
+    confirmations: Map<RequestId, HashSet<PublicKey>>,
+}
+
+impl Default for MultiSigContract {
+    fn default() -> Self {
+        env::panic(b"Multisig contract should be initialized before usage")
+    }
+}
+
+#[near_bindgen]
+impl MultiSigContract {
+    /// Initialize multisig contract.
+    /// @params num_confirmations: k of n signatures required to perform operations.
+    #[init]
+    pub fn new(num_confirmations: u32) -> Self {
```

Comment:
You may want to initialize keys here by passing them and then adding them.

Reply:
This should be done via transaction (started writing that in readme how to do that)

Reply:
The suboptimal part is that there is no way to check all access keys from contract.